janusgraph-test/src/test/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasUniquePropertyOptimizerStrategyTest.java

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize;
+
+import org.apache.tinkerpop.gremlin.process.traversal.TraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.NoOpBarrierStep;
+import org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration;
+import org.janusgraph.graphdb.tinkerpop.optimize.strategy.AdjacentVertexHasUniquePropertyOptimizerStrategy;
+import org.junit.jupiter.api.Test;
+
+import static org.janusgraph.graphdb.JanusGraphBaseTest.option;
+import static org.janusgraph.testutil.JanusGraphAssert.assertNumStep;
+import static org.janusgraph.testutil.JanusGraphAssert.assertSameResultWithOptimizations;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategyTest extends OptimizerStrategyTest {
+
+    private static TraversalStrategy optimizer = AdjacentVertexHasUniquePropertyOptimizerStrategy.instance();
+
+    @Test
+    public void testWithBackendAccess() {
+        makeSampleGraph();
+
+        // AdjacentVertexHasUniquePropertyOptimizer
+        assertNumStep(1, 0, g.V(sv[0]).out().has("uniqueId", 0), HasStep.class);
+        assertNumStep(1, 0, g.V(sv[0]).out().barrier(2500).has("uniqueId", 0), NoOpBarrierStep.class);
```

Comment:
Question (not directly related to this line but since you are testing barrier step): is there any difference if your Strategy is put before / after LazyBarrierStrategy?

Reply:
I had to ensure that barrier steps between the `out()` step and the `has()` step don't break the strategy, because I want my generated filter step to be merged into a preceeding vertex step. This should happen no matter if there is a `barrier()` in between these steps, so I tested it.